File: dypro_leetcode/dia 021/canJump.js
var canJump = function(nums) {
    let dp = {};
    function rec(index){
        if (index + nums[index] >= nums.length - 1) return true;
        if (index in dp) return dp[index];
        dp[index] = false;
        for (let i = index + 1; i < 1 + index + nums[index]; i ++){
            if (rec(i)) {
                return true;
            }
        }  
        return false;
    }
    return rec(0);
};

File: leetcode/dia 021/canJump.py
class Solution:
    def canJump(self, nums: List[int]) -> bool:
        dp = {}
        def rec(index):
            if index + nums[index] >= len(nums) - 1:
                return True 
            if index in dp:
                return dp[index]
            dp[index] = False
            for i in range(index + 1, index + 1 + nums[index]):
                if rec(i):
                    return True
            return dp[index]
        return rec(0)